crates/ty_python_semantic/src/dunder_all.rs

```diff
@@ -0,0 +1,469 @@
+    /// statement.
+    ///
+    /// Returns [`None`] if module resolution fails, invalid syntax, or if the module does not have
+    /// a `__all__` variable.
+    fn dunder_all_names_for_import_from(
+        &self,
+        import_from: &ast::StmtImportFrom,
+    ) -> Option<&'db FxHashSet<Name>> {
+        let module_name =
+            ModuleName::from_import_statement(self.db, self.file, import_from).ok()?;
+        let module = resolve_module(self.db, &module_name)?;
+        dunder_all_names(self.db, module.file())
+    }
+
+    /// Evaluate the given expression and return its truthiness.
+    ///
+    /// Returns [`None`] if the expression type doesn't implement `__bool__` correctly.
+    fn evaluate_test_expr(&self, expr: &ast::Expr) -> Option<Truthiness> {
+        infer_expression_types(self.db, self.index.expression(expr))
+            .expression_type(expr.scoped_expression_id(self.db, self.scope))
+            .try_bool(self.db)
+            .ok()
+    }
+
+    /// Add valid names to the set.
+    ///
+    /// Returns `false` if any of the names are invalid.
+    fn add_names(&mut self, exprs: &[ast::Expr]) -> bool {
+        for expr in exprs {
+            let Some(name) = create_name(expr) else {
+                return false;
+            };
+            self.names.insert(name);
+        }
+        true
+    }
+
+    /// Consumes `self` and returns the collected set of names.
+    ///
+    /// Returns [`None`] if `__all__` is not defined in the current module or if it contains
+    /// invalid elements.
+    fn into_names(self) -> Option<FxHashSet<Name>> {
+        if self.origin.is_none() {
+            None
+        } else if self.invalid {
+            tracing::debug!("Invalid `__all__` in `{}`", self.file.path(self.db));
+            None
+        } else {
+            Some(self.names)
+        }
+    }
+}
+
+impl<'db> StatementVisitor<'db> for DunderAllNamesCollector<'db> {
+    fn visit_stmt(&mut self, stmt: &'db ast::Stmt) {
+        if self.invalid {
+            return;
+        }
+
+        match stmt {
+            ast::Stmt::ImportFrom(import_from @ ast::StmtImportFrom { names, .. }) => {
+                for ast::Alias { name, asname, .. } in names {
+                    // `from module import *` where `module` is a module with a top-level `__all__`
+                    // variable that contains the "__all__" element.
+                    if name == "*" {
+                        // Here, we need to use the `dunder_all_names` query instead of the
+                        // `exported_names` query because a `*`-import does not import the
+                        // `__all__` attribute unless it is explicitly included in the `__all__` of
+                        // the module.
+                        let Some(all_names) = self.dunder_all_names_for_import_from(import_from)
+                        else {
+                            continue;
+                        };
```

Comment:
I think I'll need to mark here as `invalid` (`self.invalid = true`) as well if `dunder_all_names_for_import_from` returns `None` as otherwise it adds new false positives.

Reply:
Is there a reason you didn't do this yet? Should we at least have a TODO comment for it, or do you plan to add it in this PR?

Reply:
I'm planning to do it in this PR itself.